azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/auth/base_renewable_token_authentication_provider.py

```diff
@@ -81,14 +96,108 @@ def generate_new_sas_token(self):
             token = _device_token_format.format(quoted_resource_uri, signature, str(expiry))
 
         self.sas_token_str = str(token)
+        self._schedule_token_update(self.token_renewal_interval - self.token_renewal_margin)
+        self._notify_token_updated()
+
+    def set_token_renewal_interval(self, token_renewal_interval, token_renewal_margin):
+        """
+        Update the token renewal interval for this authorization object.
+        Thie method does not generate a new token and it does not cause the current
+        token renewal timer to restart.
+
+        Two numbers define token renewal.
+
+        The first number is the interval.  This defines the amount of time that the token
+        is valid.  The interval is encoded in the token as an offset from the current time,
+        as based on the Unix epoch.  In other words, the expiry (se=) value in the token
+        is the number of seconds after 00:00 on January 1, 1970 that the token expires.
+
+        The second number that defines the token renewal interval is the margin.  This is
+        the number of seconds before expiration that we want to generate a new token.  Since
+        the clocks on different computers can drift over time, they will all have different
+        definitions of what "current time" is, so the margin needs to be set so there is a
+        very small chance that there is no time overlap where one computer thinks the token
+        is expired and another doesn't.
+
+        When the timer is set to renew the SAS token, the timer is set for
+        (token_renewal_interval - token-renewal_margin) seconds in the future.  In this way,
+        the token will be renewed close to it's expiration time, but not so close that
+        we risk a problem caused by clock drift.
+
+        :param token_renewal_interval: The number of seconds, in the future, that
+        you want the token to expire.
+
+        :param token_renewal_margin: The number of seconds, before the token expires,
+        that you want the token renewal process to begin.
+
+        :returns: None
+        """
+        logger.info(
+            "changing renewal interval for (%s,%s) to (%d,%d)",
+            self.device_id,
+            self.module_id,
+            token_renewal_interval,
+            token_renewal_margin,
+        )
+        self.token_renewal_interval = token_renewal_interval
+        self.token_renewal_margin = token_renewal_margin
+
+    def _cancel_token_update_timer(self):
+        """
+        Cancel any future token update operations.  This is typically done as part of a
+        teardown operation
+        """
+        t = self._token_update_timer
+        self._token_update_timer = None
+        if t:
+            logger.info("Canceling token update timer for (%s,%s)", self.device_id, self.module_id)
+            t.cancel()
```

Comment:
>t.cancel() [](start = 12, length = 10)

can this raise an exception if the timer has already been cancelled? should we catch it?

Reply:
It does not throw if you cancel a timer that hasn't been started or if you cancel a timer that's already been cancelled.  Either way, the code here should protect against double-cancel.


---
In reply to: [240404496](https://github.com/Azure/azure-iot-sdk-python-preview/pull/33#discussion_r240404496) [](ancestors = 240404496)